### PR TITLE
Add built-in support for Oidc tokens

### PR DESCRIPTION
**Pull request by AlexF4Dev**
The pull request adds support for OIDC access tokens. It uses PKCE flow to acquire tokens - no stored secrets required. The tokens are cached, and refresh token is used when the access token is expired. It can be used with any identity provider which supports PKCE OpenID Connect flow.

Example for Google:
```
GET https://openidconnect.googleapis.com/v1/userinfo
Authorization: Bearer {{$oidcAccessToken clientId:786371393912-ahu1p5p81n2pj4fa90bo33pirr1mroeu.apps.googleusercontent.com callbackPort:7777 authorizeEndpoint:https://accounts.google.com/o/oauth2/v2/auth tokenEndpoint:https://oauth2.googleapis.com/token}}
Accept: application/json
```
Example for Microsoft:
```
GET https://graph.microsoft.com/oidc/userinfo
Authorization: Bearer {{$oidcAccessToken clientId:336c0138-03e7-4183-a398-3e63445dd5f5 callbackPort:7777 authorizeEndpoint:https://login.microsoftonline.com/consumers/oauth2/v2.0/authorize tokenEndpoint:https://login.microsoftonline.com/consumers/oauth2/v2.0/token}}
Accept: application/json
```

The example will call an API with google access token. For google the app should be registered as UWP to use PKCE without the client secret.

### CODE REVIEW

**Comment by Huachao** [comment]
@AlexF4Dev please check these alerts

**Comment by AlexF4Dev** [comment]
> @AlexF4Dev please check these alerts

The first alert is probably false positive, because this is not a webapp. I have fixed regex warnings.

File: src/common/constants.ts [hunk 1]
```diff
@@ -24,6 +24,11 @@ export const AzureActiveDirectoryDescription = "Prompts to sign in to Azure AD a
 export const AzureActiveDirectoryV2TokenVariableName = "$aadV2Token";
 export const AzureActiveDirectoryV2TokenDescription = "Prompts to sign in to Azure AD V2 and adds the token to the request";
 
+export const OidcVariableName = "$oidcAccessToken";
+export const OidcDescription = "Prompts to sign in to an Oidc provider  and adds the token to the request";
```
Comment (Huachao) [inline src/common/constants.ts:28]: remove extra whitespace in description

File: src/utils/memoryCache.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    public  get(key: string): T | undefined {
+        return this.cache.get(key);
+    }
+
+    public  set(key: string, value: T) {
+        this.cache.set(key, value);
+    }
+
+    public  clear(): void {
```
Comment (Huachao) [inline src/utils/memoryCache.ts:27]: remove extra whitespace between access modifier `public` and method name in this file

File: src/utils/auth/oidcClient.ts [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+          return;
+        }
+
+        const authCodeResponse = CodeLoopbackClient.getDeserializedQueryString(url);
+        if (authCodeResponse.code) {
+          const redirectUri = await this.getRedirectUri();
+          res.writeHead(302, { location: redirectUri }); // Prevent auth code from being saved in the browser history
+          res.end();
+        } else {
+          res.end(`Authorization Server Error:${JSON.stringify(authCodeResponse)}`);
```
[the remark this answers is not in the document]
Comment (Huachao) [reply src/utils/auth/oidcClient.ts:95]: @AlexF4Dev please resolve this

**Review by Huachao — CHANGES_REQUESTED** [review]

**Comment by AlexF4Dev** [comment]
Trust nobody even IDP :-). I have added sanitizeHTML to prevent the script injection. Do we need support for DPOP/MTLS?  It could be used to access API's protected by FAPI.

**Review by Huachao — APPROVED** [review]

**Comment by Huachao** [comment]
@AlexF4Dev Merged, thanks

**Comment by noeldamps** [comment]
@AlexF4Dev We handle the authorization with Keycloak. Have you maybe tried this feature with Keycloak? I only get an error that there is no valid http token. :/

**Pull request merged**